File: original/654. Maximum Binary Tree/654.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    """
        Title: 654. Maximum Binary Tree
        Time Complexity: O(N^2)
        Algorithm: 
            1. Divide & Conquer(Follow the statement, no variation)
        
        Date: 07/17/2021 14:45			
        Time: 184 ms -- beat 91.93%
        Memory: 15.2 MB  -- beat 9.75%
    """
    def dfs(self, i, j) -> TreeNode:
        if i == j:
            return TreeNode(self.nums[i])
        
        else:
            max_ = max(self.nums[i:j+1])
            mid = self.table[max_]

            root = TreeNode(max_)
            if i < mid:
                root.left = self.dfs(i, mid-1)

            if mid < j:
                root.right = self.dfs(mid+1, j)

            return root
        
    def constructMaximumBinaryTree(self, nums: List[int]) -> TreeNode:
        self.nums = nums
        self.table = {}
        for i, num in enumerate(self.nums):
            self.table[num] = i
        
        return self.dfs(0, len(self.nums)-1)
        
        
    

File: tagged/sprint/binaryTree/654. Maximum Binary Tree/654.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

#include <vector>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode* construct(vector<int>& nums, int begin, int end) {
        if (begin > end) {
            return nullptr;
        }
        
        int maxIndex = begin;
        for (int i = begin+1; i <= end; i ++) {
            if (nums[i] > nums[maxIndex]) {
                maxIndex = i;
            }
        }
        TreeNode* root = new TreeNode(nums[maxIndex]);
        root->left = construct(nums, begin, maxIndex-1);
        root->right = construct(nums, maxIndex+1, end);
        return root;
    }
    
    TreeNode* constructMaximumBinaryTree(vector<int>& nums) {
        return construct(nums, 0, nums.size()-1);
    }
};

